custom_components/rpi_gpio/switch.py

```diff
@@ -95,19 +96,21 @@ async def async_added_to_hass(self) -> None:
         else: 
             _LOGGER.debug(f"setting initial persistent state for: {self._port}. state: {state.state}")
             self._attr_is_on = True if state.state == STATE_ON else False
-        self._hub.add_switch(self, self._port, self._active_low, self._bias, self._drive_mode)
-        self.async_write_ha_state()
+            self.async_write_ha_state()
+        self._line = self._hub.add_switch(self, self._port, self._active_low, self._bias, self._drive_mode)
```

Comment:
As mentioned on the discussion it is a bad practice to fail within the entity, what happens if the port configuration is incorrect?

Reply:
In ```binary_sensor``` I was able to meet your requirement to fail early. in ```switch``` it is not possible as can configure the GPIO only when I know the desired state. If I do it earlier the switch state might toggle between the two.
Also, please note, this is not worse than what we used to have. Before this change all GPIO execution was done even later when we got the ```EVENT_HOMEASSISTANT_START``` notification

Reply:
I tried few approaches to fix that but there is no way to initialize the gpio in the ```switch.__init__``` and than later set the value in ```async_added_to_hass``` without losing its state between the two. For whatever reason the ```chip.request_lines``` will turn off the line unless specifically asked to keep it on.
Bottom line: This is not worse than what we used to have so I think it should not block this change.

Reply:
It is not allowed to do any IO operation on the class init, 
The init should happen before, you can see an example here:
https://github.com/thecode/ha-rpi_gpio/blob/config-flow/custom_components/rpi_gpio/switch.py

BTW, I highly recommend to look at this branch, it has a good implementation config flow for the `Switch` platform by @engrbm87

Reply:
Are you referring to the usage of ```hass.async_add_executor_job```? I can try to use it but I'm not sure why it is needed in our case. The ```chip.request_lines``` function is not really doing any long running operation. It is not accessing the network or storage.
My understanding is that we should not halt the main thread with long running operations and we really don't do that so I think it is adding complexity when it is not needed.

Reply:
@thecode , ping...

Reply:
Not referring to `hass.async_add_executor_job`, but how to do the IO related outside of the entity class, this way if it fails the platform will fail.
I think we should reconsider using a data update coordinator (https://github.com/thecode/ha-rpi_gpio/discussions/325#discussioncomment-11493212) or another approach, maybe even start with config flow which has great benefit for users and will give the path towards a common update coordinator.

I know that part of the issues I raised exists also in the current code, but IMO if we refactor it is better to address them to do only one refactoring, however if you prefer to continue with this PR I will finish the review.

Reply:
I made another checkin which add early validation for the switch so if it is not usable, we will fail at ```__init__``` as you requested and not in the ```async_added_to_hass``` function.
I looked at ```DataUpdateCoordinator``` and it is not relevant for our case. The doc says it is relevant for integrations which want to coordinate polling between entities. In our case we dont do poling at all as we get immediate push notification via our registration to file descriptor changes provided by the undelaying ```libgpiod``` library.
@thecode. I think you can review the code as is. It currently meets all your requirements. With regards to config flow, this should be done in separate change, and I don't suspect it will require any change in the new hub design.

Reply:
> I looked at ```DataUpdateCoordinator``` and it is not relevant for our case. The doc says it is relevant for integrations which want to coordinate polling between entities. In our case we dont do poling at all as we get immediate push 

Actually I mostly use `DataUpdateCoordinator` for push updates and I was the one who initiated support for push updates (events) in `DataUpdateCoordinator`